### PR TITLE
Fix #2306 : Fix discussions were not getting in search result

### LINKED ISSUE #2306: Subforum/Forum discussions are not showing in network search
**Describe the bug**
Subforum discussions are not showing up in network search as well as in forum instant search result. For more clarification please refer to my video: [https://www.loom.com/share/611b8cbdcaad491b86c7441036157fee ](https://www.loom.com/share/611b8cbdcaad491b86c7441036157fee )

**Support ticket links**
[https://secure.helpscout.net/conversation/1473710819/134568?folderId=4314898](https://secure.helpscout.net/conversation/1473710819/134568?folderId=4314898)

### PR DESCRIPTION
**Pull request by buntybuddyboss**
### All Submissions:

* [x] Have you followed the [Pull Requests Guidelines](https://github.com/buddyboss/buddyboss-platform/wiki/Submitting-Pull-Requests#pull-request-guidelines)?
* [x] Does your code follow the [WordPress' Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Fixes #2306 .

### How to test the changes in this Pull Request:

1. Enable forum search in network search setting ( BuddyBoss--> Settinggs--> Search ) : https://prnt.sc/12ixkyz
2. Try to search any discussion title from any Forum/Sub-Forum from search box in header : https://prnt.sc/12ixnnk
3. Hit Enter and see the result on result page : https://prnt.sc/12ixovv

### Proof Screenshots or Video

https://www.loom.com/share/5194047bc21f4572aedbb1a608760174

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

d08458c2b3effdfc8915134815b1186a1288c074 Improved search query for forums

### CODE REVIEW

**Review by KartikSuthar — DISMISSED** [review]

**Comment by mohit-buddyboss** [comment]
@buntybuddyboss 

FIXED> Tested on dev branch of the platform and also on 'fix/2306' of the platform.
What was the issue ? >Under network search individual forums discussions and their sub forum discussions were not appearing also on forum directory search under suggestions individual forums and sub forums discussions were not coming but on search result page they are appearing.
Video: https://www.loom.com/share/b82e5ad2567846e8b5e97e239db72a07

Tested this by enabling forums discussions under network search. Tested the search result for individual forum and sub forum discussion as well as group forum discussion. Also verified same on forum directory page search. Also, tested this by enabling only discussion in network search and by enabling all the options under network search.

Dev branch> on dev branch if 'Members' option is enabled under network search then in network search suggestions "members" name are appearing although on search result page correct search result appears. This issue is coming only on dev branch and not in release and 'fix/2306' branch. 
For Proof see below videos:
Search without enabling Media option>https://www.loom.com/share/d5c452ef6ef4485ea9c354756c7637b1
Search after enabling members and all other options> https://www.loom.com/share/51c72a961862433498f2c02d78aae7e5


Related issue is resolve now and PR is ready to merge. 

cc @chetansatasiya @sourabhmatolia

**Comment by sourabhmatolia** [comment]
@asaquzzamanbb 
Please take over this task.

**Comment by asaquzzamanbb** [comment]
> @asaquzzamanbb
> Please take over this task.

Ok, i will check it.

File: src/bp-search/classes/class-bp-search-bbpress-forums-topics.php [hunk 1]
```diff
@@ -127,18 +127,36 @@ function ( $group_id ) {
 				$post_status = array( "'publish'" );
 			}
 
+			$post_status = join( ',', $post_status );
+			// Get all forum regarding group.
+			$forum_query = "SELECT p.ID as forum_id 
+				FROM $wpdb->posts p 
+					LEFT JOIN $wpdb->postmeta pm ON pm.post_id = p.ID
+				WHERE 1=1
+					AND p.post_type = 'forum'
```
Comment (KartikSuthar) [inline src/bp-search/classes/class-bp-search-bbpress-forums-topics.php:136]: use bbp_get_forum_post_type()

File: src/bp-search/classes/class-bp-search-bbpress-forums-topics.php [hunk 2]
```diff
@@ -165,6 +183,30 @@ function ( $group_id ) {
...
+		 * @param int $forum_id
+		 * 
+		 * @return array
+		 */
+		public function nested_child_forum_ids( $forum_id ) {
+			global $wpdb;
+
+			// SQL query for getting all nested child forum id from parent forum id.
+			$sql = "SELECT ID
+				FROM  ( SELECT * FROM {$wpdb->posts} WHERE post_type = 'forum' AND post_status IN ( 'publish', 'private', 'hidden' ) ) forum_sorted,
```
Comment (KartikSuthar) [inline src/bp-search/classes/class-bp-search-bbpress-forums-topics.php:200]: use bbp_get_forum_post_type()

**Review by KartikSuthar — CHANGES_REQUESTED** [review]

File: src/bp-search/classes/class-bp-search-bbpress-forums-topics.php (hunk 1 above)
Comment (asaquzzamanbb) [reply src/bp-search/classes/class-bp-search-bbpress-forums-topics.php]: Done here https://github.com/buddyboss/buddyboss-platform/pull/2379/commits/f354ee6f1759d69af38e9872405fbcf4461639da

File: src/bp-search/classes/class-bp-search-bbpress-forums-topics.php (hunk 2 above)
Comment (asaquzzamanbb) [reply src/bp-search/classes/class-bp-search-bbpress-forums-topics.php]: Done here https://github.com/buddyboss/buddyboss-platform/pull/2379/commits/f354ee6f1759d69af38e9872405fbcf4461639da

**Comment by asaquzzamanbb** [comment]
@KartikSuthar 

This PR has updated with your requirement. Could you please review this PR again.

**Review by chetansatasiya — DISMISSED** [review]

**Comment by chetansatasiya** [comment]
@asaquzzamanbb merged to dev and add in testing

**Comment by asaquzzamanbb** [comment]
@chetansatasiya 

Merge done to dev branch. https://github.com/buddyboss/buddyboss-platform/commit/94caa072956fddd4ccee56bbfa95851adc738a55

**Comment by mohit-buddyboss** [comment]
@asaquzzamanbb 

Tested on the dev branch of the platform as well as on 'fix/2306' of the platform.

> **ISSUE :** Private Group forum discussions and Private Individual forum discussions are appearing under network and forum search suggestions and on search result page.
> Video:> https://www.loom.com/share/cf77060a029040efb5014e9b2a7b3b56

@asaquzzamanbb Above issue got resolved but it created issue again which we fixed. 

**ISSUE:** Individual Public Forum discussion and Sub  Forum discussion are not coming network search suggestion as well on search result page. Also under forum search suggestions they are not appearing.

Video: https://www.loom.com/share/f38ef636b05c46d392c4925e581b28b4

cc @chetansatasiya @sourabhmatolia

**Comment by asaquzzamanbb** [comment]
@chetansatasiya @KartikSuthar 

Could you please review this PR again.

File: src/bp-search/classes/class-bp-search-bbpress-forums-topics.php [hunk 3]
```diff
@@ -165,6 +206,32 @@ function ( $group_id ) {
 			);
 		}
 
+		/**
+		 * Get all nested child forum ids.
+		 * 
+		 * @since BuddyBoss 1.6.2
```
Comment (chetansatasiya) [inline src/bp-search/classes/class-bp-search-bbpress-forums-topics.php:212]:
```suggestion
		 * @since BuddyBoss 1.6.3
```

**Review by chetansatasiya — COMMENTED** [review]
LGTM change the since tag and good to go

**Review by chetansatasiya — APPROVED** [review]

**Comment by asaquzzamanbb** [comment]
@chetansatasiya 

Merge done! https://github.com/buddyboss/buddyboss-platform/commit/30fd561f21f80fea17dffdef01bd2968848ee0f3

**Comment by mohit-buddyboss** [comment]
@asaquzzamanbb 

Tested on dev branch of the platform.

Issue> Users don't have access to Individual Private forum discussion(not associated with any group) but getting suggestions and result in 'Network search' and only suggestions not result in 'Forum Search' 
Video: https://www.loom.com/share/fecfbe38730e4986a478a9da18fa32e0

cc @sourabhmatolia @chetansatasiya

**Comment by mohit-buddyboss** [comment]
@asaquzzamanbb 

FIXED>Tested on dev branch of the platform. 

1. Tested for Public group discussion, Public individual forum discussion , Public sub forum discussion under network and forum search. Search is working fine. 
Video: https://www.loom.com/share/0216df9a225f459092ab0725f3026907

2. Tested for Private and Hidden group discussion, Private and Hidden individual forum discussion under network and forum search. Search is working fine. 

Video: https://www.loom.com/share/9e643a801be84ccabc05cce58f3c6150

Related issue is resolve now and PR is ready for release.

cc @chetansatasiya @sourabhmatolia

**Pull request merged**